### ISSUE TITLE
`torch.cumsum` drops the sign of IEEE `-0.0` on CPU (CUDA and NumPy preserve it)

### ISSUE THREAD

**Issue by lv20031001** [pytorch/pytorch#195851]
### 🐛 Describe the bug

`torch.cumsum` on CPU turns IEEE-754 negative zero (`-0.0`) into positive zero. The input has the sign bit set; after `cumsum` it does not. NumPy and PyTorch CUDA keep the sign; `torch.cumprod` on the same CPU tensor also keeps it.

This is a signed-zero semantics issue, not a rounding / tolerance difference under the numerical accuracy notes: `(-0.0) + (-0.0)` is `-0.0` in IEEE-754. `torch.signbit` is the right check here.

Related but different polarity or op (CPU keeps `-0.0`, CUDA drops it, or a different function): #162235, #114569, #181801, #87131. I could not find a `cumsum`-specific report of CPU dropping `-0.0`.

```python
import torch

x = torch.tensor([-0.0, -0.0, 1.0])
print(torch.signbit(x).tolist())                   # [True, True, False]
print(torch.signbit(torch.cumsum(x, 0)).tolist())  # [False, False, False]
```

Expected: CPU `cumsum` preserves `-0.0`, so the second line should also be `[True, True, False]`.

Cross-checks on the same input, torch 2.14.0+cpu:

```python
import numpy as np
np.signbit(np.cumsum(np.array([-0.0, -0.0, 1.0])))  # [True, True, False]
torch.signbit(torch.cumprod(x, 0)).tolist()         # [True, False, False]
```

Same CPU loss on 2.13.0+cu130 and 2.8.0+cu128; CUDA on those builds gives `[True, True, False]`. Also reproduces for `float16` / `bfloat16` / `float32` / `float64`. `sum([-0.0])` is `+0.0` on both devices, so this is specific to `cumsum`.

### Versions

CPU arm: PyTorch 2.14.0+cpu, NumPy 2.5.2, Python 3.12.3, Ubuntu 22.04.5 (x86_64)
CUDA arm: PyTorch 2.13.0+cu130, RTX 5090, driver 580.95.05
Also reproduces on 2.8.0+cu128 (original discovery run).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by devnp2007** [comment]
Hi @lv20031001, I'd like to investigate this issue and work on a fix if it hasn't already been picked up. I'll first reproduce the behavior locally and trace the CPU `cumsum` implementation to understand where the signed-zero information is being lost, then add a focused regression test and propose the smallest fix. Please let me know if there's any existing work or preferred direction I should be aware of.

**Comment by abhinav-phi** [comment]
Hi! I'd like to work on this.

Plan: I'll investigate the reported behavior, implement the change/fix per the issue description, and add tests to cover it.

Could you please assign this issue to me? Thanks!

/assign take

**Comment by lv20031001** [comment]
Thanks both! No existing work from my side — feel free to investigate.
@devnp2007 commented first with a clear plan, so I'd suggest coordinating here to avoid duplicate PRs.

**Comment by devnp2007** [comment]
Thanks @lv20031001, and thanks @abhinav-phi for being interested in this as well.

I'll proceed with the investigation I outlined above, starting with reproducing the issue and tracing the CPU `cumsum` implementation to identify the exact source of the signed-zero behavior. I'll share my findings here before making a PR so we can coordinate and avoid duplicate work.

@abhinav-phi happy to coordinate if you find anything useful during your investigation as well.

**Comment by abhinav-phi** [comment]
Thanks @devnp2007 — let's coordinate here as @lv20031001 suggested; I won't open a duplicate PR.

One hypothesis worth checking first, since it matches the exact output in the report: if the CPU scan kernel initializes the accumulator to `0` (positive) and then does `acc = acc + x[i]`, then in round-to-nearest mode `(+0.0) + (-0.0) = +0.0`, so leading `-0.0` inputs collapse to `+0.0` — which is exactly the `[False, False, False]` result seen for `[-0.0, -0.0, 1.0]`. NumPy's `add.accumulate` seeds the accumulator from the first element, which is why it preserves the sign. (`cumprod` survives any init because multiplication propagates the sign of zero.)

If that's right, the fix may be as small as seeding the accumulator from the first element (or special-casing a zero prefix) in the CPU `cumsum`/scan kernel, plus `signbit`-based regression tests across the dtypes listed in the report.

I'll verify the hypothesis on HEAD and post what I find here before touching any PR, so @devnp2007 can confirm or rule it out.

**Comment by devnp2007** [comment]
Thanks @abhinav-phi — this matches what I found as well. I traced the CPU implementation to cpu_cum_base_kernel / cumsum_cpu_kernel in ReduceOpsKernel.cpp. The CPU cumsum initializes the accumulator with +0 and then performs cum_number += self_data[i], so a leading -0.0 becomes +0.0 on the first iteration. The accumulator uses acc_type, so the issue isn't related to reduced precision.

I also confirmed the CUDA path preserves the first element through its inclusive scan behavior. My current fix is to seed the CPU cumsum accumulator from the first input element and then perform the existing accumulation from the second element onward, with a focused signbit regression test covering float16, bfloat16, float32, and float64.

I'll proceed with validating this locally before opening the PR.

**Comment by devnp2007** [comment]
I opened PR #196206 with a fix for this issue. The fix initializes the CPU cumsum accumulator from the first input element so IEEE-754 signed zero is preserved. I also added a regression test covering float16, bfloat16, float32, and float64.

Thanks @abhinav-phi for coordinating on the investigation.
